internal/parseutil/filereader.go

```diff
@@ -0,0 +1,56 @@
+package parseutil
+
+import (
+	"bytes"
+	"io"
+)
+
+// FileReader serves two purposes:
+// 1. read input sources and join them for single source
+// 2. track the lines counts of each input source
+type FileReader struct {
+	files  []file
+	source bytes.Buffer
+}
+
+type file struct {
+	path      string
+	startLine int // the number of first line of this file in joined source
+	lines     int
+}
+
+// AddFile adds input source
+func (fr *FileReader) AddFile(path string, source io.Reader) error {
+	curLen := fr.source.Len()
+	_, err := fr.source.ReadFrom(source)
+	if err != nil {
+		return err
+	}
+	if !bytes.HasSuffix(fr.source.Bytes(), []byte("\n")) {
+		fr.source.WriteByte('\n')
+	}
+	content := fr.source.Bytes()[curLen:]
+	lines := bytes.Count(content, []byte("\n"))
+	startLine := 1
+	if len(fr.files) > 0 {
+		lastFile := fr.files[len(fr.files)-1]
+		startLine = lastFile.startLine + lastFile.lines
+	}
+	fr.files = append(fr.files, file{path, startLine, lines})
+	return nil
+}
+
+// FileLine resolves global line number in joined source to a local line number in a file (identified by path)
```

Comment:
Let's wrap the comment lines to <80 chars, and update the wording a little here:

```go
// FileLine resolves an overall line number from the concatenated source code
// to the local line number in that source file (identified by path).
```